diff -Naurp pcp-6.3.7.orig/src/python/pmda.c pcp-6.3.7/src/python/pmda.c
--- pcp-6.3.7.orig/src/python/pmda.c	2025-06-26 19:39:47.002695985 +1000
+++ pcp-6.3.7/src/python/pmda.c	2025-06-26 19:40:16.426718001 +1000
@@ -417,11 +417,11 @@ refresh_all_clusters(int numclusters, in
 	PyList_SET_ITEM(list, i, num);
     }
 
-    arglist = Py_BuildValue("(N)", list);
+    arglist = Py_BuildValue("(O)", list);
+    Py_DECREF(list);
     if (arglist == NULL)
 	return -ENOMEM;
     result = PyObject_Call(refresh_all_func, arglist, NULL);
-    Py_DECREF(list);
     Py_DECREF(arglist);
     if (result == NULL)
 	return callback_error("refresh_all_clusters");
