diff -Naurp pcp-6.3.1.orig/configure pcp-6.3.1/configure
--- pcp-6.3.1.orig/configure	2024-09-11 11:45:29.000000000 +1000
+++ pcp-6.3.1/configure	2024-10-02 11:50:38.938538019 +1000
@@ -5442,11 +5442,11 @@ if test x$ac_prog_cxx_stdcxx = xno
 then :
   { printf "%s\n" "$as_me:${as_lineno-$LINENO}: checking for $CXX option to enable C++11 features" >&5
 printf %s "checking for $CXX option to enable C++11 features... " >&6; }
-if test ${ac_cv_prog_cxx_11+y}
+if test ${ac_cv_prog_cxx_cxx11+y}
 then :
   printf %s "(cached) " >&6
 else $as_nop
-  ac_cv_prog_cxx_11=no
+  ac_cv_prog_cxx_cxx11=no
 ac_save_CXX=$CXX
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
@@ -5488,11 +5488,11 @@ if test x$ac_prog_cxx_stdcxx = xno
 then :
   { printf "%s\n" "$as_me:${as_lineno-$LINENO}: checking for $CXX option to enable C++98 features" >&5
 printf %s "checking for $CXX option to enable C++98 features... " >&6; }
-if test ${ac_cv_prog_cxx_98+y}
+if test ${ac_cv_prog_cxx_cxx98+y}
 then :
   printf %s "(cached) " >&6
 else $as_nop
-  ac_cv_prog_cxx_98=no
+  ac_cv_prog_cxx_cxx98=no
 ac_save_CXX=$CXX
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
@@ -9457,7 +9457,7 @@ printf %s "checking if Perl is Y2038-saf
 			    #
 			    { printf "%s\n" "$as_me:${as_lineno-$LINENO}: result: no" >&5
 printf "%s\n" "no" >&6; }
-			    have_perl=false
+			    #have_perl=false -- el9 pre-existing-breakage
 			    ;;
 		    67768036191676799)
 			    # all good
diff -Naurp pcp-6.3.1.orig/configure.ac pcp-6.3.1/configure.ac
--- pcp-6.3.1.orig/configure.ac	2024-09-11 11:45:29.000000000 +1000
+++ pcp-6.3.1/configure.ac	2024-10-02 11:50:35.841529740 +1000
@@ -1401,7 +1401,7 @@ AS_IF([test "x$do_perl" != "xno"], [
 			    # can't play in this sandpit
 			    #
 			    AC_MSG_RESULT([no])
-			    have_perl=false
+			    #have_perl=false -- el9 pre-existing-breakage
 			    ;;
 		    67768036191676799)
 			    # all good
