diff -Naurp pcp-6.3.7.orig/configure pcp-6.3.7/configure
--- pcp-6.3.7.orig/configure	2025-03-31 12:41:26.000000000 +1100
+++ pcp-6.3.7/configure	2025-03-31 14:00:04.534091532 +1100
@@ -9618,7 +9618,7 @@ printf %s "checking if Perl is Y2038-saf
 			    #
 			    { printf "%s\n" "$as_me:${as_lineno-$LINENO}: result: no" >&5
 printf "%s\n" "no" >&6; }
-			    have_perl=false
+			    #have_perl=false -- el9 pre-existing-breakage
 			    ;;
 		    67768036191676799)
 			    # all good
diff -Naurp pcp-6.3.7.orig/configure.ac pcp-6.3.7/configure.ac
--- pcp-6.3.7.orig/configure.ac	2025-03-31 12:41:26.000000000 +1100
+++ pcp-6.3.7/configure.ac	2025-03-31 13:59:38.095025279 +1100
@@ -1406,7 +1406,7 @@ AS_IF([test "x$do_perl" != "xno"], [
 			    # can't play in this sandpit
 			    #
 			    AC_MSG_RESULT([no])
-			    have_perl=false
+			    #have_perl=false -- el9 pre-existing-breakage
 			    ;;
 		    67768036191676799)
 			    # all good
