diff -Naurp pcp-6.2.2.orig/tmpfiles.init.setup.in pcp-6.2.2/tmpfiles.init.setup.in
--- pcp-6.2.2.orig/tmpfiles.init.setup.in	2024-05-07 10:37:04.000000000 +1000
+++ pcp-6.2.2/tmpfiles.init.setup.in	2024-05-16 09:18:49.200282601 +1000
@@ -3,7 +3,7 @@
 # Type	Path			Mode	User	Group	Age	Argument
 D	PCP_RUN_DIR		0775	PCP_USER	PCP_GROUP	-	-
 d	PCP_LOG_DIR		0775	PCP_USER	PCP_GROUP	-	-
-d	PCP_LOG_DIR/pmcd	0775	root	root	-	-
+d	PCP_LOG_DIR/pmcd	0755	root	root	-	-
 d	PCP_LOG_DIR/pmfind	0775	PCP_USER	PCP_GROUP	-	-
 d	PCP_LOG_DIR/pmie	0775	PCP_USER	PCP_GROUP	-	-
 d	PCP_LOG_DIR/pmlogger	0775	PCP_USER	PCP_GROUP	-	-
