commit 23bfdcbac0b0fb7bd33f092c6f2ad56889480335
Author: Andreas Gerstmayr <agerstmayr@redhat.com>
Date:   Thu Oct 27 14:22:37 2022 +0200

    build: use vendored vmlinux.h when compiling bpftool

diff --git a/vendor/GNUmakefile b/vendor/GNUmakefile
index a0c29d4b45..57151e4671 100644
--- a/vendor/GNUmakefile
+++ b/vendor/GNUmakefile
@@ -3,7 +3,7 @@ include $(TOPDIR)/src/include/builddefs
 
 default_pcp default:
 ifeq "$(PMDA_BPF)" "true"
-	$(MAKE) -C github.com/libbpf/bpftool/src
+	$(MAKE) -C github.com/libbpf/bpftool/src VMLINUX_H=$(PMDABPF_VMLINUXH)
 endif
 
 install_pcp install:
