diff -Naurp pcp-6.3.7.orig/GNUmakefile pcp-6.3.7/GNUmakefile
--- pcp-6.3.7.orig/GNUmakefile	2025-03-27 17:01:59.000000000 +1100
+++ pcp-6.3.7/GNUmakefile	2025-06-26 19:23:24.874469186 +1000
@@ -135,6 +135,8 @@ pcp.lsm src/include/builddefs src/includ
 
 tmpfiles.init.setup:	tmpfiles.init.setup.in
 	sed < $< > $@ \
+	    -e "s@PCP_VAR_DIR@$(PCP_VAR_DIR)@" \
+	    -e "s@PCP_TMP_DIR@$(PCP_TMP_DIR)@" \
 	    -e "s@PCP_RUN_DIR@$(PCP_RUN_DIR)@" \
 	    -e "s@PCP_LOG_DIR@$(PCP_LOG_DIR)@" \
 	    -e "s/PCP_GROUP/$(PCP_GROUP)/" \
diff -Naurp pcp-6.3.7.orig/tmpfiles.init.setup.in pcp-6.3.7/tmpfiles.init.setup.in
--- pcp-6.3.7.orig/tmpfiles.init.setup.in	2024-07-30 16:43:55.000000000 +1000
+++ pcp-6.3.7/tmpfiles.init.setup.in	2025-06-26 19:23:24.875070062 +1000
@@ -1,10 +1,16 @@
 # tmpfiles.d(5) config for Performance Co-Pilot startup components
 #
-# Type	Path			Mode	User	Group	Age	Argument
-D	PCP_RUN_DIR		0775	PCP_USER	PCP_GROUP	-	-
-d	PCP_LOG_DIR		0775	PCP_USER	PCP_GROUP	-	-
-d	PCP_LOG_DIR/pmcd	0755	root	root	-	-
-d	PCP_LOG_DIR/pmfind	0775	PCP_USER	PCP_GROUP	-	-
-d	PCP_LOG_DIR/pmie	0775	PCP_USER	PCP_GROUP	-	-
-d	PCP_LOG_DIR/pmlogger	0775	PCP_USER	PCP_GROUP	-	-
-d	PCP_LOG_DIR/pmproxy	0775	PCP_USER	PCP_GROUP	-	-
+# Path                          Mode User Group            Age Argument
+D PCP_RUN_DIR                      0775 PCP_USER  PCP_GROUP              -   -
+d PCP_VAR_DIR                  0755 root root             -   -
+d PCP_VAR_DIR/config           0755 root root             -   -
+d PCP_VAR_DIR/config/pmda      0775 PCP_USER  PCP_GROUP              -   -
+d PCP_VAR_DIR/config/pmie      0775 PCP_USER  PCP_GROUP              -   -
+d PCP_VAR_DIR/config/pmlogger  0775 PCP_USER  PCP_GROUP              -   -
+d PCP_TMP_DIR              0775 PCP_USER  PCP_GROUP              -   -
+d PCP_LOG_DIR                  0775 PCP_USER  PCP_GROUP              -   -
+d PCP_LOG_DIR/pmcd             0755 root root             -   -
+d PCP_LOG_DIR/pmfind           0775 PCP_USER  PCP_GROUP              -   -
+d PCP_LOG_DIR/pmie             0775 PCP_USER  PCP_GROUP              -   -
+d PCP_LOG_DIR/pmlogger         0775 PCP_USER  PCP_GROUP              -   -
+d PCP_LOG_DIR/pmproxy          0775 PCP_USER  PCP_GROUP              -   -
